src/libcore/iter.rs

```diff
@@ -1573,18 +1793,94 @@ impl<'a, I: DoubleEndedIterator + ?Sized> DoubleEndedIterator for &'a mut I {
     fn next_back(&mut self) -> Option<I::Item> { (**self).next_back() }
 }
 
-/// An iterator that knows its exact length
+/// An iterator that knows its exact length.
+///
+/// Many [`Iterator`]s don't know how many times they will iterate, but some do.
+/// If an iterator knows how many times it can iterate, providing access to
+/// that information can be useful. For example, if you want to iterate
+/// backwards, a good start is to know where the end is.
+///
+/// Furthermore, sometimes you can perform optimizations if you know exactly
+/// how much iteration you're going to be doing, such as pre-allocating memory
+/// of the right size.
```

Comment:
This is actually generally done with the lower bound of `size_hint` (e.g. all iterators can benefit, not just exact ones). One of the main reasons to add this iterator was to have something like: `foo.enumerate().rev()` work because if you call `next_back` you'll need to know how long the entire collection is to return the index of the last element.